aten/src/ATen/native/TensorTransformations.cpp

```diff
@@ -57,6 +57,38 @@ Tensor flip_cpu(const Tensor& self, IntList dims) {
   return out_tensor;
 }
 
+Tensor roll_cpu(const Tensor& self, IntList shifts, IntList dims) {
+  // todo: support rolling along no or multiple dimensions as in numpy.roll.
+  AT_CHECK(dims.size() == 1, "only single dimension roll currently supported");
+  AT_CHECK(shifts.size() == dims.size(), "shifts and dimensions must align");
+  // If the first dimension is zero, this is an empty tensor and rolls do nothing.
+  // Return a clone so the caller can safely modify result, and avoid a div by
+  // zero error below.
+  if( self.size(0) == 0 ) {
+    return self.clone();
+  }
+  int64_t dim = dims[0];
+  int64_t size = self.size(dim);
+  int64_t start = (size - shifts[0]) % size;
+  // Behavior of % is different in C++ vs Python for negative numbers. This
+  // corrects the difference.
+  if( start < 0 ) start = start + size;
+
+  auto tensors = self.unbind(dim);
+  std::vector<Tensor> vec = std::vector<Tensor>(size);
+  int64_t index = 0;
+  for (int64_t i = start; i < size; i++) {
+    vec[index++] = tensors[i];
+  }
+
+  for (int64_t i = 0; i < start; i++) {
+    vec[index++] = tensors[i];
+  }
+
+  auto stacked = at::stack(vec, dim);
```

Comment:
This code does worse than numpy on CPU regardless of tensor size, worse on GPU for small tensors, but performs better for large tensors on GPU. I guess this makes sense since I wasn't really trying to optimize the CPU code, and smaller tensors will benefit less from parallelism on the GPU.

Large:
>>> timeit.timeit("torch.roll(x,5,1)",setup='import torch;x=torch.randn(10, 100, 10, 100, device="cuda")',number=10000)
1.1422520764172077
>>> timeit.timeit("torch.roll(x,5,1)",setup='import torch;x=torch.randn(10, 100, 10, 100)',number=10000)
71.12064150907099
>>> timeit.timeit("numpy.roll(x,5,1)",setup='import numpy;x=numpy.random.randn(10, 100, 10, 100)',number=10000)
6.136326801031828

Small:
>>> timeit.timeit("torch.roll(x,1,0)", setup='import torch;x=torch.arange(10, device="cuda")',number=100000)
8.392850634176284
>>> timeit.timeit("torch.roll(x,1,0)", setup='import torch;x=torch.arange(10)', number=100000)
28.554980190936476
>>> timeit.timeit("numpy.roll(x,1,0)", setup='import numpy;x=numpy.arange(10)', number=100000)
0.9924737298861146

Reply:
Interesting; so 10x-30x slower on CPU.

Reply:
I was going to need to make significant performance improvements, but just re-compiling without debug symbols gets this a lot closer to numpy for CPU performance. Now cuda is also faster than cpu for small tensors as well.

>>> import timeit
>>> timeit.timeit("torch.roll(x,5,1)",setup='import torch;x=torch.randn(10, 100, 10, 100, device="cuda")',number=10000)
0.4632282108068466
>>> timeit.timeit("torch.roll(x,5,1)",setup='import torch;x=torch.randn(10, 100, 10, 100)',number=10000)
13.862711012829095
>>> timeit.timeit("numpy.roll(x,5,1)",setup='import numpy;x=numpy.random.randn(10, 100, 10, 100)',number=10000)
5.864686117041856
>>>
>>> timeit.timeit("torch.roll(x,1,0)", setup='import torch;x=torch.arange(10, device="cuda")',number=100000)
1.0122237978503108
>>> timeit.timeit("torch.roll(x,1,0)", setup='import torch;x=torch.arange(10)', number=100000)
1.341067599132657
>>> timeit.timeit("numpy.roll(x,1,0)", setup='import numpy;x=numpy.arange(10)', number=100000)
1.241553642321378
>>>